optimum/onnxruntime/modeling_ort.py

```diff
@@ -0,0 +1,689 @@
+    Args:
+        input_ids (`torch.Tensor` of shape `({0})`):
+            Indices of input sequence tokens in the vocabulary.
+            Indices can be obtained using [`AutoTokenizer`](https://huggingface.co/docs/transformers/autoclass_tutorial#autotokenizer). 
+            See [`PreTrainedTokenizer.encode`](https://huggingface.co/docs/transformers/main_classes/tokenizer#transformers.PreTrainedTokenizerBase.encode) and
+            [`PreTrainedTokenizer.__call__`](https://huggingface.co/docs/transformers/main_classes/tokenizer#transformers.PreTrainedTokenizerBase.__call__) for details.
+            [What are input IDs?](https://huggingface.co/docs/transformers/glossary#input-ids)
+        attention_mask (`torch.Tensor` of shape `({0})`, *optional*):
+            Mask to avoid performing attention on padding token indices. Mask values selected in `[0, 1]`:
+            - 1 for tokens that are **not masked**,
+            - 0 for tokens that are **masked**.
+            [What are attention masks?](https://huggingface.co/docs/transformers/glossary#attention-mask)
+        token_type_ids (`torch.Tensor` of shape `({0})`, *optional*):
+            Segment token indices to indicate first and second portions of the inputs. Indices are selected in `[0, 1]`:
+            - 1 for tokens that are **sentence A**,
+            - 0 for tokens that are **sentence B**.
+            [What are token type IDs?](https://huggingface.co/docs/transformers/glossary#token-type-ids)
+"""
+
+
+@add_start_docstrings(
+    """
+    Base ORTModel class for implementing models using ONNX Runtime. The ORTModel implements generic methods for interacting
+    with the Hugging Face Hub as well as exporting vanilla transformers models to ONNX using `transformers.onnx` toolchain.
+    The ORTModel implements additionally generic methods for optimizing and quantizing Onnx models.
+    """,
+)
+class ORTModel(OptimizedModel):
+    base_model_prefix = "onnx_model"
+
+    def __init__(self, model=None, config=None, **kwargs):
+        self.model = model
+        self.config = config
+        self.model_save_dir = kwargs.get("model_save_dir", None)
+        self.latest_model_name = kwargs.get("latest_model_name", "model.onnx")
+
+    def forward(self, *args, **kwargs):
+        raise NotImplementedError
+
+    @staticmethod
+    def load_model(path: Union[str, Path], provider=None):
+        """
+        loads ONNX Inference session with Provider. Default Provider is if GPU available else `CPUExecutionProvider`
+        Arguments:
+            path (:obj:`str` or :obj:`Path`):
+                Directory from which to load
+            provider(:obj:`str`):
+                Onnxruntime provider to use for loading the model, defaults to `CUDAExecutionProvider` if GPU is
+                available else `CPUExecutionProvider`
+        """
+        if provider is None:
+            provider = "CUDAExecutionProvider" if _is_gpu_available() else "CPUExecutionProvider"
```

Comment:
What about wrapping thoses names within an enum to make it more user-friendly for the user?

```python
class ORTDevice(Enum):
    CPU = "CPUExecutionProvider"
    CUDA = "CUDAExecutionProvider"
    TENSORRT = "TensorRTExecutionProvider"
    ...
```

I find those name not very friendly for someone coming from PyTorch/Device

or even something a la PyTorch:

```python
class device:
    def __init__(self, specs: str):
        parse_specs(specs)
    
   @property
    def is_cpu(self) -> bool:
        pass
  
   @property
   def is_cuda(self) -> bool:
      pass

   @property
   def num_cores(self) -> Optional[int]:
        return specs["core"]  # for symplicity

   @property
   def gpu_id(self) -> Optional[int]:
        return specs["gpu_id"]
   
   def to_execution_provider_specs(self) -> Tuple[str, dict[str, any]]
       return "CPUExecutionProvider", {"num_intraops_threads": self.num_cores}

   
cpu_bound_device = device("cpu:0-16")
cpu_cores_device = device("cpu:16")
gpu_device = device("cuda:0")
```

Reply:
Great Idea and suggestion. But let's try to get this PR in. I would create separate small issues we can then work on after the PR is merged.